components/builder-db/src/pool.rs

```diff
@@ -60,6 +95,59 @@ impl Pool {
             thread::sleep(Duration::from_millis(connection_retry_ms));
         }
     }
+
+    pub fn get_raw(&self)
+                   -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let conn = self.inner.get().map_err(Error::ConnectionTimeout)?;
+        Ok(conn)
+    }
+
+    pub fn get_shard
+        (&self,
+         shard_id: u32)
+         -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let conn = self.inner.get().map_err(Error::ConnectionTimeout)?;
+        debug!("Switching to shard {}", shard_id);
+
+        let schema_name = if self.testing {
+            format!("test_{}_shard_{}", self.test_number, shard_id)
+        } else {
+            format!("shard_{}", shard_id)
+        };
+        let sql_search_path = format!("SET search_path TO {}", schema_name);
+        conn.execute(&sql_search_path, &[]).map_err(Error::SchemaSwitch)?;
+        Ok(conn)
+    }
+
+    pub fn get<T: Routable>
+        (&self,
+         routable: &T)
+         -> Result<r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>> {
+        let mut hasher = FnvHasher::default();
```

Comment:
It'd be much better if we could get this on the connection itself instead of allocating on every `get()`